proto/api/v1beta1/ensign.proto

```diff
@@ -37,33 +37,57 @@ service Ensign {
     rpc TopicNames(PageInfo) returns (TopicNamesPage) {}
     rpc TopicExists(TopicName) returns (TopicExistsInfo) {}
 
-    // Info provides some statistical information about the state of a project
+    // Info provides statistics and metrics describing the state of a project
     rpc Info(InfoRequest) returns (ProjectInfo) {}
 
     // Implements a client-side heartbeat that can also be used by monitoring tools.
     rpc Status(HealthCheck) returns (ServiceState) {}
 }
 
-// Publication messages are sent back to publishers from the server. Generally they are
-// responses to receiving events (e.g. ack and nack) but the last message contains
+// PublisherRequest messages are sent from the publisher to the server. Generally they
+// are events that need to be published but the first message must be a stream
+// initialization message to ensure that the publisher is authenticated and can identify
+// itself. This message can be extended to allow for other interactions between the
+// publisher and the server including topic queries, redirects, etc.
+message PublisherRequest {
+    oneof embed {
+        EventWrapper event = 1;
+        OpenStream open_stream = 2;
+    }
+}
+
+// PublisherReply messages are sent back to publishers from the server. Generally they
+// are responses to receiving events (e.g. ack and nack) but the last message contains
 // information about the performance of the publisher and the topic itself.
-message Publication {
+// This message can be extended to allow more general interactions between the publisher
+// and the server including redirects, topic queries, etc.
+message PublisherReply {
```

Comment:
So here we could potentially send back the topic mappings to help the SDKs?

Reply:
Yes - any general interaction in the pub/sub stream could be wrapped in these messages.